db/dcrpg/queries.go

```diff
@@ -3619,6 +3620,182 @@ func appendBlockFees(charts *cache.ChartData, rows *sql.Rows) error {
+	blocks := charts.Blocks
+	for rows.Next() {
+		var blockHeight uint64
+		var totalMixed int64
+		if err := rows.Scan(&blockHeight, &totalMixed); err != nil {
+			log.Errorf("Unable to scan for MixedVoutsPerBlock fields: %v", err)
+			return err
+		}
+		if totalMixed < 0 {
+			totalMixed *= -1
+		}
+
+		// Converting to atoms.
+		blocks.TotalMixed = append(blocks.TotalMixed, uint64(totalMixed))
+	}
+	return rows.Err()
+}
+
+// retrieveAnonymitySet retrieves any block total mixed data that is newer than the data
+// in the provided ChartData. This data is used to plot anonymity-set on the /charts page.
+// This is the Fetcher half of a pair that make up a cache.ChartUpdater.
+func retrieveAnonymitySet(ctx context.Context, db *sql.DB, _ *cache.ChartData) (*sql.Rows, error) {
+	rows, err := db.QueryContext(ctx, internal.SelectMixedVouts)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+// Append the result from retrieveAnonymitySet to the provided ChartData. This
+// is the Appender half of a pair that make up a cache.ChartUpdater.
+func appendAnonymitySet(charts *cache.ChartData, rows *sql.Rows) (err error) {
+	var vals, fundHeights, spendHeights []int64
+
+	var maxHeight int64
+	minHeight := int64(math.MaxInt64)
+
+
+	for rows.Next() {
+		var value, fundHeight, spendHeight int64
+		var spendHeightNull sql.NullInt64
+		var tree uint8
+		err = rows.Scan(&value, &fundHeight, &spendHeightNull, &tree)
+		if err != nil {
+			return err
+		}
+
+		if spendHeightNull.Valid {
+			spendHeight = spendHeightNull.Int64
+		} else {
+			spendHeight = -1
+		}
+
+		if fundHeight < minHeight {
+			minHeight = fundHeight
+		}
+
+		if spendHeight > maxHeight {
+			maxHeight = spendHeight
+		}
+
+		vals = append(vals, value)
+		fundHeights = append(fundHeights, fundHeight)
+		spendHeights = append(spendHeights, spendHeight)
+	}
+
+	err = rows.Err()
+	if err != nil {
+		log.Errorf("Unable to scan for BlockCoinJoins fields: %v", err)
+	}
+
+	blocks := charts.Blocks
+
+	// block anonymity set
+	for _, h := range blocks.Height[len(blocks.AnonymitySet) :] {
+		var anonymitySet int64
+		for iu := range vals {
+			if h >= uint64(fundHeights[iu]) && (h < uint64(spendHeights[iu]) || spendHeights[iu] == -1) {
+				anonymitySet += vals[iu]
+			}
+		}
+		blocks.AnonymitySet = append(blocks.AnonymitySet, uint64(anonymitySet))
+	}
+
+	// append days
+	days := charts.Days
+
+	// Get the current first and last midnight stamps.
+	end := midnight(blocks.Time[len(blocks.Time)-1])
+	var start uint64
+	if len(days.Time) > 0 {
+		// Begin the scan at the beginning of the next day. The stamps in the Time
+		// set are the midnight that starts the day.
+		start = days.Time[len(days.Time)-1] + aDay
+	} else {
+		// Start from the beginning.
+		// Already checked for empty blocks above.
+		start = midnight(blocks.Time[0])
+	}
+
+	// Find the index that begins new data.
+	offset := 0
+	for i, t := range blocks.Time {
+		if t > start {
+			offset = i
+			break
+		}
+	}
+
```

Comment:
Why the rest of this function being done here instead of in `(*ChartData).Lengthen`, from which all this days code is taken?  Please do it in `(*ChartData).Lengthen` using  the appropriate ChartUints methods.  If that is not possible, we should create appropriate data types to allow all the day binning to be done in one spot, Lengthen.